draft-ietf-jsonpath-base.md

```diff
@@ -1402,91 +1453,125 @@ According to {{filter-selector}}, a `function-expr` is valid as a `filter-path`
+A type is a set of instances.
+Declared types enable checking a JSONPath query for well-typedness
+independent of any argument the JSONPath query is applied to.
 
 {{tbl-types}} defines the available types in terms of abstract instances, where `n` denotes a node, `v` denotes a value, and `nl` denotes
-a non-empty nodelist. The table also lists the (immediate) subtypes of each type.
-
-| Type                      | Abstract Instances                       | Subtypes                                       |
-| :--                       | :----------------                        | :------                                        |
-| `OptionalNodeOrValueType` | `Node(n)`, `Value(v)`, `Nothing`         | `OptionalNodeType`, `OptionalValueType`        |
-| `OptionalNodeType`        | `Node(n)`, `Nothing`                     |                                                |
-| `OptionalValueType`       | `Value(v)`, `Nothing`                    | `OptionalNodeType`, `ValueType`, `OptionalBooleanType` |
-| `ValueType`               | `Value(v)`                               | `BooleanType`                                  |
-| `OptionalBooleanType`     | `Value(true)`, `Value(false)`, `Nothing` | `BooleanType`                                  |
-| `BooleanType`             | `Value(true)`, `Value(false)`            |                                                |
-| `OptionalNodesType`       | `Nodes(nl)`, `Nothing`                   | `OptionalNodeType`                             |
+a nodelist.
+
+| Type                 | Abstract Instances                       |
+| :--                  | :----------------                        |
+| `ValueType`          | `Value(v)`, `Nothing`                    |
+| `LogicalType`        | `LogicalTrue`, `LogicalFalse`            |
+| `NodesType`       | `Nodes(nl)`                              |
 {: #tbl-types title="Function extension type system"}
 
 Notes:
 
-* `OptionalNodeOrValueType` is an abstraction of a `comparable` (which may appear on either side of a comparison or as a function argument).
-* `OptionalNodeType` is an abstraction of a Singular Path.
-* `ValueType` is an abstraction of a primitive value.
-* `BooleanType` is an abstraction of a primitive value that is either
-  `true` or `false`.
-* `OptionalValueType` is an abstraction of a primitive value that may
-  alternatively be absent (`Nothing`).
-* `OptionalNodesType` is an abstraction of a `filter-path` (which appears
+* `ValueType` is an abstraction of a JSON value or `Nothing`.
+* `LogicalType` is an abstraction of the result of a `logical-expr`.
+  Its two instances, `LogicalTrue` and `LogicalFalse`, are not related to
+  the JSON literals `true` and `false` and have no syntactical representation in JSONPath.
+* `NodesType` is an abstraction of a `filter-path` (which appears
   in a test expression or as a function argument).
+  Members of `NodesType` have no syntactical representation in JSONPath.
```

Comment:
Post-merge query:

@cabo Similarly, why isn't a query that always returns an empty nodelist (e.g. `$[?1 == 0]`) a valid syntactical representation of `Nothing` (a member of `NodesType` )?

Reply:
https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/403#discussion_r1122634114

Reply:
https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/413